public/javascripts/AchievementTracker.js

```diff
@@ -0,0 +1,214 @@
+                } else {
+                    badgeHtmlElement.classList.remove(BADGE_NOT_YET_EARNED_CLASS_NAME);
+                }
+
+                // The parent element 'achievements-badge-grid' starts out invisible to make initial rendering cleaner.
+                badgeHtmlElement.parentElement.style.visibility = 'visible';
+            }
+
+            let badgeEncouragementHtmlId = badgeType + '-badge-encouragement';
+            document.getElementById(badgeEncouragementHtmlId).innerHTML = this.getBadgeEncouragementHtml(badgeType, curValue);
+        }
+    }
+
+    /**
+     * Get a dynamic encouragement statement for the given badge type with the current value.
+     *
+     * @param badgeType: is one of four BadgeTypes: "missions", "distance", "labels", or "validations"
+     * @param curValue: value corresponds to the current number of completed missions, total distance, num of labels, etc.
+     *               corresponding to the passed badgeType
+     */
+    getBadgeEncouragementHtml(badgeType, curValue) {
+        // Find next badge to unlock.
+        let mapLevelsToBadge = this.mapBadges[badgeType]
+        let nextBadgeToUnlock = null;
+        if (badgeType in this.mapBadges) {
+            let mapLevelsToBadge = this.mapBadges[badgeType]
+
+            for (let level of Object.keys(mapLevelsToBadge)) {
+                let badge = mapLevelsToBadge[level];
+                if (badge.threshold > curValue) {
+                    nextBadgeToUnlock = badge;
+                    break;
+                }
+            }
+        }
+
+        let htmlStatement = '';
+
+        if (nextBadgeToUnlock != null) {
+            let diffValue = nextBadgeToUnlock.threshold - curValue;
+            let curBadge = null;
+            const curBadgeLevel = nextBadgeToUnlock.level - 1;
+            let fractionComplete = 0;
+
+            // Figure out how close they are to the next badge.
+            if (curBadgeLevel in mapLevelsToBadge) {
+                curBadge = mapLevelsToBadge[curBadgeLevel];
+                const badgeValueRange = nextBadgeToUnlock.threshold - curBadge.threshold;
+                fractionComplete = (curValue - curBadge.threshold) / badgeValueRange;
+            } else {
+                fractionComplete = curValue / nextBadgeToUnlock.threshold;
+            }
+
+            // Add an encouraging statement based on how close they are to the next badge level.
+            if (fractionComplete > 0.95) {
+                htmlStatement += i18next.t('dashboard:so-close') + ' ' + i18next.t('dashboard:just') + ' ';
+            } else if (fractionComplete > 0.85) {
+                htmlStatement += i18next.t('dashboard:wow-almost-there') + ' ' + i18next.t('dashboard:just') + ' ';
+            } else if (fractionComplete > 0.1 || curBadgeLevel > 0) {
+                let randStatement = encouragingStatements[Math.floor(Math.random() * encouragingStatements.length)];
```

Comment:
I shortened this by choosing randomly from a list instead of having the 10 if cases.